### PR TITLE
ui(topology): return early on missing data

### PR DESCRIPTION
**Pull request by mikemorris**
This probably shouldn't be merged in its current state, but is documenting a chain of runtime errors that happen when Prometheus returns empty responses to a query (not sure if its because there's no load on the service, a query is malformed, or something else...)

### CODE REVIEW

File: ui-v2/app/components/topology-metrics/series/index.js [hunk 1]
```diff
@@ -41,37 +41,39 @@ export default Component.extend({
...
     svg.selectAll('path').remove();
     svg.selectAll('rect').remove();
 
-    let bb = svg.node().getBoundingClientRect()
+    let bb = svg.node().getBoundingClientRect();
     let w = bb.width;
     let h = bb.height;
 
-    if (!this.data.series) {
+    if (!this.data.series || this.data.series.length === 0) {
```
Comment (mikemorris) [inline ui-v2/app/components/topology-metrics/series/index.js:64]: `!this.data.series` will only check if the field is undefined, added a guard against the empty array I was getting.

File: ui-v2/public/assets/metrics-providers/prometheus.js [hunk 2]
```diff
@@ -209,439 +208,452 @@
...
+        return data;
+      });
 
       // Fetch the metrics async, and return a promise to the result.
-      return all
+      return all;
     },
 
     reformatSeries: function(response) {
+      if (!response.data.result || response.data.result.length === 0) return [];
```
Comment (mikemorris) [inline ui-v2/public/assets/metrics-providers/prometheus.js:276]: To guard against the responses like this I was getting from Prometheus:
```
http://localhost:4200/v1/internal/ui/metrics-proxy/api/v1/query_range?query=sum(irate(envoy_tcp_downstream_cx_rx_bytes_total%7Blocal_cluster%3D%22api%22%2C%20envoy_tcp_prefix%3D%22public_listener_tcp%22%7D%5B10m%5D))&start=1602184178.57&end=1602185078.57&step=10s&timeout=8s

{"status":"success","data":{"resultType":"matrix","result":[]}}
```
```
http://localhost:4200/v1/internal/ui/metrics-proxy/api/v1/query?query=label_replace(sum(rate(envoy_cluster_upstream_cx_connect_fail%7Bconsul_service%3D%22api%22%7D%5B15m%5D))%20by%20(local_cluster)%2C%20%22downstream%22%2C%20%22%241%22%2C%20%22local_cluster%22%2C%20%22(.*)%22)&time=1602185078.57

{"status":"success","data":{"resultType":"matrix","result":[]}}
```

File: ui-v2/app/components/topology-metrics/series/index.js [hunk 3]
```diff
@@ -169,46 +173,51 @@ export default Component.extend({
...
+      .text(niceTimeWithSeconds(mouseTime));
 
     tipVals.nodes().forEach((n, i) => {
-      let val = stackData[i][tipIdx][1] - stackData[i][tipIdx][0];
-      select(n)
-        .text(this.formatTooltip(keys[i], val));
+      let tipData = stackData[i][tipIdx];
+      if (!tipData) return;
+
+      let val = tipData[1] - tipData[0];
```
Comment (mikemorris) [inline ui-v2/app/components/topology-metrics/series/index.js:192]: This is needed to guard against the `[]` return from `reformatSeries`, without that early return though (letting it throw an error instead) I think `stackData` might've been entirely undefined...

**Comment by mikemorris** [comment]
Ugh, Prettier added a lot of formatting changes - added comments for the three actual spots I changed.

File: ui-v2/app/components/topology-metrics/series/index.js (hunk 3 above)
Comment (banks) [reply ui-v2/app/components/topology-metrics/series/index.js:192]: Doesn't the fix you made on line 64 prevent this ever happening. I think if there is no series data we should never get this far. I'll look again.

File: ui-v2/app/components/topology-metrics/series/index.js [hunk 4]
```diff
@@ -147,20 +151,20 @@ export default Component.extend({
...
+        tooltip.style('visibility', 'visible');
+        cursor.style('visibility', 'visible');
         // We update here since we might redraw the graph with user's cursor
         // stationary over it. If that happens mouseover fires but not
         // mousemove but the tooltip and cursor are wrong (based on old data).
         self.updateTooltip(event, data, stackData, keys, x, tooltip, tipVals, cursor);
       })
-      .on("mousemove", function(d, i){
+      .on('mousemove', function(d, i) {
         self.updateTooltip(event, data, stackData, keys, x, tooltip, tipVals, cursor);
```
Comment (mikemorris) [inline ui-v2/app/components/topology-metrics/series/index.js:163]: @banks Re https://github.com/hashicorp/consul/pull/8886/files#r502369139, `updateTooltip` appears to get registered here on `mouseover` and `mousemove` events (which can fire quite rapidly, might need to be throttled) regardless of series data validity (had to comment out these two lines initially when I had everything broken).

**Pull request closed without merging**